src/python/pants/core/goals/publish.py

```diff
@@ -131,6 +133,7 @@ class PublishPackages:
     """
 
     names: tuple[str, ...]
+    packages: tuple[BuiltPackage, ...]
```

Comment:
Not to keen on adding this field here. It's only use is by the "internal" nature handled by the `package_for_publish` rule to propagate to the goal rule. Rather, I'd suggest adding a new result type for that rule that returns all the packages as well as the `PublishProcesses`.

That way we don't need to touch the backends rules with this change.